src/RangeSlider.elm

```diff
@@ -0,0 +1,349 @@
+        , div
+            [ Html.Attributes.class "input-range-label input-range-label--current-value" ]
+            [ Html.text <| valueAttributes.formatter { value = valueAttributes.value, max = commonAttributes.max } ]
+        , div
+            [ Html.Attributes.class "input-range-label" ]
+            [ Html.text <| commonAttributes.maxFormatter { value = commonAttributes.max } ]
+        ]
+
+
+
+-- API
+
+
+defaultFormatter : { value : Float } -> String
+defaultFormatter value =
+    String.fromFloat value.value
+
+
+defaultValueFormatter : { value : Float, max : Float } -> String
+defaultValueFormatter values =
+    if values.value == values.max then
+        ""
+
+    else
+        String.fromFloat values.value
+
+
+initSingleSlider :
+    { min : Float
+    , max : Float
+    , step : Float
+    , value : Float
+    , onChange : String -> a
+    , onInput : String -> a
+    , onClick : String -> a
+    , valueFormatter : { value : Float, max : Float } -> String
+    , minFormatter : { value : Float } -> String
+    , maxFormatter : { value : Float } -> String
+    }
+    -> Slider a
+initSingleSlider attrs =
+    SingleSlider
+        { commonAttributes =
+            { min = attrs.min
+            , max = attrs.max
+            , step = attrs.step
+            , click = attrs.onClick
+            , minFormatter = attrs.minFormatter
+            , maxFormatter = attrs.maxFormatter
+            }
+        , valueAttributes =
+            { value = attrs.value
+            , change = attrs.onChange
+            , input = attrs.onInput
+            , formatter = attrs.valueFormatter
+            }
+        }
+
+
+initDoubleSlider :
+    { min : Float
+    , max : Float
+    , step : Float
+    , lowValue : Float
+    , highValue : Float
+    , onChange : String -> a
+    , onInput : String -> a
+    , onClick : String -> a
+    , valueFormatter : { value : Float, max : Float } -> String
+    , minFormatter : { value : Float } -> String
+    , maxFormatter : { value : Float } -> String
+    }
+    -> Slider a
+initDoubleSlider attrs =
+    DoubleSlider
+        { commonAttributes =
+            { min = attrs.min
+            , max = attrs.max
+            , step = attrs.step
+            , click = attrs.onClick
+            , minFormatter = attrs.minFormatter
+            , maxFormatter = attrs.maxFormatter
+            }
+        , lowValueAttributes =
+            { value = attrs.lowValue
+            , change = attrs.onChange
+            , input = attrs.onInput
+            , formatter = attrs.valueFormatter
+            }
+        , highValueAttributes =
+            { value = attrs.highValue
+            , change = attrs.onChange
+            , input = attrs.onInput
+            , formatter = attrs.valueFormatter
+            }
+        }
+
+
+updateValue :
+    { value : Maybe Float
+    , lowValue : Maybe Float
+    , highValue : Maybe Float
```

Comment:
Can't think of a better way to handle this - ideas|?

Reply:
As a quick improvement, I would split the `updateValue` into 3 different functions, one for each value since we only really update one at a time, right? But yeah, we do end up with functions that only work specifically for one type of the slider which is ok-ish.

Alternatively, using using two different opaque types for the sliders would ensure only the right functions can be called with the right type, and most of the common logic is still reused via internal functions.

Reply:
An idea I had was that instead of using two different opaque types, it may be possible to use a "phantom type" (not sure what to call this thing). As in:
```elm
type Single = Single
type Double = Double

type Slider t msg = SingleSlider {...} | DoubleSlider {...} -- no need to use `t` in the actual value, it is only for differentiating which functions can be used

initSingle : SingleConfig msg -> Slider Single msg
updateValue : Float -> Slider Single msg -> Slider Single msg -- force it to only work on Single Slider, you can do the same for Double
view : Slider t msg -> Html msg -- no need to change this function it should just work
```

But now that I am thinking about it again and looking at the code the only function that actually takes "advantage" of "not caring" about if it is a Single or Double slider is the `view`, so it may actually be better to just do two separate opaque types and create two `view` functions.

```elm
type SingleSlider msg = SingleSlider { ... }
type DoubleSlider msg = DoubleSlider { ... }

singleInit : SingleConfig msg -> SingleSlider msg
singleUpdate : Float -> SingleSlider msg -> SingleSlider msg
singleView : SingleSlider msg -> Html msg -- you would need a `doubleView` and the other double functions
```

Hopefully not too confusing. Not sure what looks better or what is better for flexibility here. 😅 
Maybe the "phantom type" is more flexible since you can create functions that act on both types and create functions that act on only one type. With the double opaque types, you always need to two functions, one for each type, but with good naming it doesn't look bad as well. I will let you experiment and decide on one. 😛

Reply:
I've gone for using two opaque types to define the sliders.  After playing about with "Phantom Types" for a bit, I came to the conclusion that it wouldn't offer many benefits as opposed to using two separate Opaque Types. 

We would still have to have two functions for each slider (e.g `singleUpdate` and `doubleUpdate`) and we would still have to pattern match within the function to exclude the opposiute one -  because we haven't "tagged" the `SingleSlider` and `DoubleSlider` values with specific types - we've just limited the arguments to the function.